### ISSUE TITLE
prefer-typeof-undefined rule proposal

### ISSUE THREAD

**Issue by AndrewLeedham** [eslint/eslint#12754]
**Please describe what the rule should do:**
This would prefer the use of `typeof x === 'undefined'` over `x === undefined`. This is somewhat possible with `no-undefined` however this rule also stops object values etc being initialised with `undefined`. Perhaps this could just be a setting for `no-undefined` or a rule in its own right?

**What category of rule is this? (place an "X" next to just one item)**

[x] Warns about a potential error (problem)
[ ] Suggests an alternate way of doing something (suggestion)
[x] Enforces code style (layout)
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

Examples of incorrect code for this rule:
```js
/*eslint prefer-typeof-undef: "error"*/
if (foo === undefined) {
  bar();
}

while (qux != undefined) {
  baz();
}
```

Examples of correct code for this rule:
```js
/*eslint prefer-typeof-undef: "error"*/
if (typeof foo === 'undefined') {
  bar();
}

while (typeof qux != 'undefined') {
  baz();
}
```

**Why should this rule be included in ESLint (instead of a plugin)?**
`undefined` can be reassigned and thus checking against its literal value could cause a false-positive. This is a core JS concept and overlaps with the `no-undefined` rule, so seems like it should be included here.

**Are you willing to submit a pull request to implement this rule?**
Yes

**Comment by mdjermanovic** [comment]
Hi @AndrewLeedham, thanks for the rule proposal!

Just to clarify the issue, did you mean [no-undefined](https://eslint.org/docs/rules/no-undefined) instead of [no-undef](https://eslint.org/docs/rules/no-undef)?

I'm guessing `no-undefined` but had to ask because `no-undef` rule has [typeof](https://eslint.org/docs/rules/no-undef#typeof) option, so it's also related to `typeof` and `undefined` in a certain way.

**Comment by AndrewLeedham** [comment]
Hi @mdjermanovic 

Good catch. I meant `no-undefined`, I'll update the proposal.

**Comment by mdjermanovic** [comment]
Thanks!

> undefined can be reassigned and thus checking against its literal value could cause a false-positive. This is a core JS concept and overlaps with the no-undefined rule, so seems like it should be included here.

This is reasonable. The issue with overwriting global `undefined` may no longer exist in most environments, but the variable can be also shadowed which could cause the same problem.

However, this applies to any use of the global `undefined` variable, so I'm not sure what would be a particular reason to disallow just `=== undefined` (and `!==`, `==`, `!=`) instead of any use, which is already possible with the `no-undefined` rule.

**Comment by AndrewLeedham** [comment]
I guess the reasoning there would be a stylistic choice. So a project can ensure that all the undefined checks are consistent. The `undefined` being shadowed or redefined in my mind is just a justification for preferring `typeof` in the first place.

**Comment by mdjermanovic** [comment]
> I guess the reasoning there would be a stylistic choice. So a project can ensure that all the undefined checks are consistent. The `undefined` being shadowed or redefined in my mind is just a justification for preferring `typeof` in the first place.

Thanks for the clarification! I think it's reasonable as a stylistic choice for consistency.

Also, `typeof` doesn't throw if the variable hasn't been declared, so as a side effect this might also catch some errors like checking for the existence of a global variable with `if (SomeGlobal !== undefined)`.

A new rule requires a champion and 3 :+1: from other members to be accepted, per the [process](https://eslint.org/docs/maintainer-guide/issues#accepting-issues).

**Comment by kaicataldo** [comment]
> Also, typeof doesn't throw if the variable hasn't been declared, so as a side effect this might also catch some errors like checking for the existence of a global variable with if (SomeGlobal !== undefined).

This is the most compelling argument for me, though I'm not sure this meets the high bar we now have to set for new rules. I think this can also be enforced using [`no-restricted-syntax`](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50ICBcIm5vLXJlc3RyaWN0ZWQtc3ludGF4XCI6IFtcbiAgICBcImVycm9yXCIsXG4gICAge1xuICAgICAgICBcInNlbGVjdG9yXCI6IFwiQmluYXJ5RXhwcmVzc2lvbltvcGVyYXRvcj0nPT09J11bcmlnaHQudHlwZT0nSWRlbnRpZmllciddW3JpZ2h0Lm5hbWU9J3VuZGVmaW5lZCddXCIsXG4gICAgICAgIFwibWVzc2FnZVwiOiBcIlVzZSB0eXBlb2YgeCA9PT0gJ3VuZGVmaW5lZCcgaW5zdGVhZC5cIlxuICAgIH0sXG4gICAgICB7XG4gICAgICAgIFwic2VsZWN0b3JcIjogXCJCaW5hcnlFeHByZXNzaW9uW29wZXJhdG9yPSchPT0nXVtyaWdodC50eXBlPSdJZGVudGlmaWVyJ11bcmlnaHQubmFtZT0ndW5kZWZpbmVkJ11cIixcbiAgICAgICAgXCJtZXNzYWdlXCI6IFwiVXNlIHR5cGVvZiB4ICE9PSAndW5kZWZpbmVkJyBpbnN0ZWFkLlwiXG4gICAgfVxuXSAqL1xuXG5mb28gPT09IHVuZGVmaW5lZDtcbmZvbyAhPT0gdW5kZWZpbmVkOyIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6MTEsInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnsianN4Ijp0cnVlfX0sInJ1bGVzIjp7fSwiZW52Ijp7ImJyb3dzZXIiOmZhbHNlfX19).

**Comment by g-plane** [comment]
I think we can implement this rule in a separated 3rd-party plugin.
